ACE/ace/Malloc_T.cpp

```diff
@@ -177,6 +177,469 @@ ACE_Dynamic_Cached_Allocator<ACE_LOCK>::free (void * ptr)
     this->free_list_.add ((ACE_Cached_Mem_Pool_Node<char> *) ptr);
 }
 
+template <class ACE_LOCK>
+ACE_Cascaded_Dynamic_Cached_Allocator<ACE_LOCK>::ACE_Cascaded_Dynamic_Cached_Allocator
+  (size_t initial_n_chunks, size_t chunk_size)
+    : initial_n_chunks_ (initial_n_chunks),
+      chunk_size_ (chunk_size),
+      chunk_sum_ (0)
+{
+  ACE_ASSERT (this->chunk_size_ > 0);
+
+  comb_alloc_ptr tmp;
+  // If ACE_NEW fails, the hierarchy_ will be reconstructed when malloc API is called.
+  ACE_NEW (tmp, comb_alloc_type(this->initial_n_chunks_, this->chunk_size_));
+
+  // Increase the chunk sum if succeed.
+  this->chunk_sum_ += tmp->pool_depth();
+  this->hierarchy_.push_back(tmp);
+}
+
+template <class ACE_LOCK>
+ACE_Cascaded_Dynamic_Cached_Allocator<ACE_LOCK>::~ACE_Cascaded_Dynamic_Cached_Allocator ()
+{
+  for (size_t h = 0; h < this->hierarchy_.size(); h++)
+  {
+    delete this->hierarchy_[h];
+  }
+
+  this->hierarchy_.clear();
+}
+
+template <class ACE_LOCK> void *
+ACE_Cascaded_Dynamic_Cached_Allocator<ACE_LOCK>::malloc (size_t nbytes)
+{
+  // Check if size requested fits within pre-determined size.
+  if (nbytes > this->chunk_size_)
+    return nullptr;
+
+  ACE_MT (ACE_GUARD_RETURN (ACE_LOCK, ace_mon, this->mutex_, nullptr));
+
+  void * ptr = nullptr;
+
+  for (size_t h = 0; h < this->hierarchy_.size(); h++)
+  {
+    ptr = this->hierarchy_[h]->malloc(nbytes);
+    if(ptr != nullptr)
+      break;
+  }
+
+  if(ptr == nullptr)
+  {
+    comb_alloc_ptr tmp;
+    ACE_NEW_RETURN (tmp, comb_alloc_type(this->initial_n_chunks_ * 2 * this->hierarchy_.size(), this->chunk_size_), nullptr);
+
+    // Increase the chunk sum if succeed.
+    this->chunk_sum_ += tmp->pool_depth();
+    this->hierarchy_.push_back(tmp);
```

Comment:
push_back can throw

Reply:
The `push_back` call should not encounter exceptions , because the vector has been resized before this statement. 

If the runtime has exception relative to memory,  the `resize` call will trigger the exception early.

But the memory has alloced before will leak if no exception guarantee added when exception occurs for calling APIs of this allocator

Reply:
I'm sorry, I see the exception for call of vector's resize or push_back API.

There exists leaks and bugs !

I have fixed the leaks and bugs for exception safety by adding judgement of the effect of these calls as they have strong exception safety guarantee